### PR TITLE
feat: add support for delta package uploads using go-octodiff

### PR DESCRIPTION
**Pull request by borland**
# Overview

Adds support for delta compression to the CLI.

Precursors:
- Creation of the [go-octodiff library](https://github.com/OctopusDeploy/go-octodiff), ported from our existing [C# OctoDiff](https://github.com/OctopusDeploy/OctoDiff)
- Integration of go-octodiff into the [go-octopusdeploy SDK client library for Octopus Deploy](https://github.com/OctopusDeploy/go-octopusdeploy/pull/190)

Delta compression is off by default. You can enable it with the `--use-delta-compression` flag, which has aliases of `--delta` and `-d`. For example

```
octopus package upload --package SomePackage.2.0.0.zip --use-delta-compression
```

Delta compression works by only transmitting the difference between two files - for example if the Octopus Server already holds version `1.0.0` of a nuget or zip package, and you are uploading `1.2.0` - if the zip contains a lot of the same files, then OctoDiff can avoid having to re-send these over a slow network link.

It is off by default because it is not a good fit for:
- Fast network links, e.g. physical servers connected locally, or cloud servers. The additional Disk I/O and CPU processing required to compute the delta between the old and new files can exceed the cost of simply sending the file
- Packages where there are no files in common between the old and new versions.
- `tar.gz` files. Zip files compress each contained file individually, so the same file will compress to the same output, enabling OctoDiff to detect this. Tar.Gz files compress all files together in a single unit; the compressed form of a file can change depending on what other files are also in the archive.
- Zip files that only contain other compressed files (e.g. you put a .msi in a zip). This produces the same outcome as `tar.gz` files.

If you are not sure whether it will help you, please experiment using the CLI using the files and network conditions that you have. The CLI will output messages such as these:

```
Delta upload for package MyPackage.2.0.0.nupkg. Delta size was 54.7% of full file, saving 200 MiB
Full upload for package MyPackage.2.0.0.nupkg. Delta size was 99% of full file (too large)
```

You can use an external utility such as `time` on unix or `Measure-Command` in powershell to determine whether delta compression will speed things up for you.

# Quality

During the creation of go-octodiff, I [added a significant number of tests](https://github.com/OctopusDeploy/Octodiff/commit/ff83814479eaad5a997782cc89cc67847a945271) to C# OctoDiff, to capture its behaviour at appropriate levels of detail. These tests were used to scaffold the go implementation, ensuring it produces bit-for-bit identical output across many test cases.

Additionally, a lot of manual testing has been done, over a variety of real-world inputs. I ran the C# and Go implementations side by side and verified they produced bit-for-bit identical output in all cases.

### Beta Disclaimer

I have confidence that the implementation in the CLI is of high quality, however, the introduction of any complex feature brings risk, and there is always the slim chance that a bug exists which has not been caught during the development process. As such, please consider this feature in **beta** until the end of 2024.

# Performance

_note_: Octodiff comprises 3 operations: 
- Creating a signature file for a prior version of a package
- Using the signature to build a delta against a new version of the package
- Applying the delta to reconstruct the new version

Of these, Creating the signature and applying the Delta are done by the Octopus Server and are not relevant to the CLI.
Time taken to build delta's for different file sizes on the same PC (Windows 11)

| Scenario | C# | go |
| --------- | -- | -- |
| 33 MB nuget package | 0.43s | 0.31s |
| 108 MB powershell zip | 2.3s | 2.2s |
| 2.1 GB docker image | 25.6s | 21.1s |

### CODE REVIEW

File: go.mod
Comment (borland) [inline go.mod:27]: Once I get an approval on Octodiff itself, I'll create a 1.0.0 release/tag for it in github and roll those references through to go-octopusdeploy and the CLI itself. I'll make sure this is done before merging

File: go.mod
Comment (borland) [inline go.mod:63]: Temporary just to get it to compile locally.
Once the octodiff and go-octopusdeploy changes are merged, there will be proper versions of those things and I'll remove this. I promise to do that before merging

File: pkg/cmd/package/upload/upload.go
Comment (benPearce1) [inline pkg/cmd/package/upload/upload.go:74]: can we add a note here to say the delta compression support is beta or experimental or something like that

**Review by benPearce1 — CHANGES_REQUESTED** [review]
Would like to see a callout that delta-compression is not production ready

**Comment by eero-dev** [comment]
Any chances that this will be implemented? Currently blocking us from switching from the .NET CLI

**Comment by borland** [comment]
Hi @eero-dev 

I've had a brief chat about this internally, and while I can't promise any dates, I am planning to pick this back up again next week if I am able to.

If possible, would you conser [creating an issue here](https://github.com/OctopusDeploy/cli/issues) requesting octodiff support, and add whatever information you feel comfortable sharing about your use-cases? (e.g. which kinds of files you upload, how big, etc)? This will help us justify the feature going forward, and help me target some tests to better help you.

Thanks, Orion

**Comment by borland** [comment]
Hi @eero-dev
End of year has been busy so I haven't been able to look at this. In the absence of any feedback I'll keep it at a lower priority and hope to pick up in early 2024

**Comment by eero-dev** [comment]
> Hi @eero-dev End of year has been busy so I haven't been able to look at this. In the absence of any feedback I'll keep it at a lower priority and hope to pick up in early 2024

We are using the CLI action in github to push packages to octopus whenever a new release is created, however the upload takes 10 minutes compared to 10 seconds before with .NET Cli

File: pkg/cmd/package/upload/upload.go
Comment (benPearce1) [inline pkg/cmd/package/upload/upload.go:56]: if the default is true, should this be a negative boolean flag?
Not sure if I am a fan of the negative, but just a thought

**Review by benPearce1 — APPROVED** [review]
Successfully tested this locally with a large file

first package version
```
❯ octopus package upload -p pkg1.1.0.0.zip --use-delta-compression true
Uploaded package pkg1.1.0.0.zip
    Full upload for package pkg1.1.0.0.zip. No previous versions available
    Timing: Signature 57.88ms, Upload 823.01ms
```

second package with ~100 kb difference
```
❯ octopus package upload -p pkg1.2.0.0.zip --use-delta-compression true
Uploaded package pkg1.2.0.0.zip
    Delta upload for package pkg1.2.0.0.zip.
    Delta size was 0.0% of full file, saving 252737468 bytes
    Timing: Signature 669.13ms, Build Delta 1.44s, Upload 493.55ms
```

File: pkg/cmd/package/upload/upload.go
Comment (benPearce1) [inline pkg/cmd/package/upload/upload.go:223]: I like this output, would it be more useable with human readable numbers?
e.g in my testing I got `Delta size was 0.0% of full file, saving 252737468 bytes`

**Comment by eero-dev** [comment]
Any updates on this? github is deprecating node16 actions and our version will stop working and CI time will increase like crazy 😔

**Comment by borland** [comment]
> Any updates on this? github is deprecating node16 actions and our version will stop working and CI time will increase like crazy 

Hi @eero-dev 

It's not clear what you're referring to with the node16 action?

I'm guessing you might be using our other [OctopusDeploy/install-octopus-cli-action](https://github.com/OctopusDeploy/install-octopus-cli-action) to install the legacy CLI that supports OctoDiff?

That action has been updated to support node 20, but if you're pinning an old version of it in your yaml file you may not be aware of this. Your GitHub action yaml file should be referencing v3 of the action, e.g.

```
uses: OctopusDeploy/install-octopus-cli-action@v3
```

Unfortunately I can't provide any further update on adding OctoDiff support to the Go CLI at this time

**Comment by eero-dev** [comment]
> > Any updates on this? github is deprecating node16 actions and our version will stop working and CI time will increase like crazy
> 
> Hi @eero-dev
> 
> It's not clear what you're referring to with the node16 action?
> 
> I'm guessing you might be using our other [OctopusDeploy/install-octopus-cli-action](https://github.com/OctopusDeploy/install-octopus-cli-action) to install the legacy CLI that supports OctoDiff?
> 
> That action has been updated to support node 20, but if you're pinning an old version of it in your yaml file you may not be aware of this. Your GitHub action yaml file should be referencing v3 of the action, e.g.
> 
> ```
> uses: OctopusDeploy/install-octopus-cli-action@v3
> ```
> 
> Unfortunately I can't provide any further update on adding OctoDiff support to the Go CLI at this time



> > Any updates on this? github is deprecating node16 actions and our version will stop working and CI time will increase like crazy
> 
> Hi @eero-dev
> 
> It's not clear what you're referring to with the node16 action?
> 
> I'm guessing you might be using our other [OctopusDeploy/install-octopus-cli-action](https://github.com/OctopusDeploy/install-octopus-cli-action) to install the legacy CLI that supports OctoDiff?
> 
> That action has been updated to support node 20, but if you're pinning an old version of it in your yaml file you may not be aware of this. Your GitHub action yaml file should be referencing v3 of the action, e.g.
> 
> ```
> uses: OctopusDeploy/install-octopus-cli-action@v3
> ```
> 
> Unfortunately I can't provide any further update on adding OctoDiff support to the Go CLI at this time

Yes we are unable to switch over to a newer version as it would install the Go CLI (as stated here https://github.com/OctopusDeploy/install-octopus-cli-action/blob/main/migration-guide.md) and the upload times without octodiff are way too high.

**Comment by borland** [comment]
@eero-dev ah, my apologies. I missed the fact that v3 installed the new CLI. I’ll ask internally if there are any other options

**Comment by danzhik** [comment]
Hello everyone. We have just recently updated from old .NET CLI to a new one, and after finally resolving all migration issues we almost come to a conclusion to revert back to Octo CLI. The reason is that a part of our CI/CD pipeline is creating a really weighty (hunderds of MBs) NuGet in GH actions, which we push to our Octopus server instance, and now with the new CLI, because of delta upload functionality being unavailable, uploading of this NuGet to Octopus takes about 10min vs previous 10 seconds in Octo CLI. Any progress on when we could expect the parity on this one? Thank you!

File: go.mod
Comment (borland) [inline go.mod:8]: NOSHIP: This makes it pick up the [unmerged go-octopusdeploy from my PR](https://github.com/OctopusDeploy/go-octopusdeploy/pull/190/files) so we can check we get a green build.

Needs to be replaced with a real version once the go-octopusdeploy PR merges and there's been a release

File: pkg/cmd/package/upload/upload.go
Comment (borland) [reply pkg/cmd/package/upload/upload.go]: Done

File: pkg/cmd/package/upload/upload.go
Comment (borland) [reply pkg/cmd/package/upload/upload.go]: I've changed it to be off by default, so changes to a bool now and is more clear.

File: pkg/cmd/package/upload/upload.go
Comment (borland) [reply pkg/cmd/package/upload/upload.go]: I've wrapped it with the existing `humanReadableBytes` helper we have in the CLI

**Comment by borland** [comment]
Hi @eero-dev @danzhik 

I've just [released version 2.4.0 of the CLI](https://github.com/OctopusDeploy/cli/releases/tag/v2.4.0), which adds delta compression. It is off by default, but you can enable it if you specify the `--use-delta-compression` flag according to the description on this PR.

I hope you find it useful!

**Comment by danzhik** [comment]
Great news @borland, thanks a lot! I hope to test it soon, but it seems like releases.json file still has v2.2.0 stated as the latest available, thus it's not possible to install the new version of cli using a GH action☹️

**Comment by borland** [comment]
Thanks @danzhik. I will look into releases.json sometime this week, hopefully tomorrow

**Comment by borland** [comment]
Hi @danzhik. With an assist from @IsaacCalligeros95, the releases.json should now contain the 2.4.0 version, enabling you to pull it via [install-octopus-cli-action](https://github.com/OctopusDeploy/install-octopus-cli-action). Let me know if there are any further issues, thanks!

**Comment by danzhik** [comment]
Hi @borland, that's wonderful, thanks a lot! I've been really looking forward to it since v2.3.0. After a sample GH actions run, I'm happy to report that feature works as intended, huge thanks for a prompt turnover on this one! Not sure if it matches the performance of the old one for 100%, but even if not, the discrepancy is really minor and in our case can be neglected. Again, thanks for the efforts!

**Pull request merged**
